File: python/agent_vfs/operations.py
from __future__ import annotations

import re
import uuid

from .db.types import Database, NodeRow
from .errors import (
    EditConflictError,
    FileSizeError,
    IsDirectoryError,
    NotDirectoryError,
    NotFoundError,
)
from .paths import all_ancestors, base_name, glob_to_like, normalize, parent_path

DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB


class FileSystem:
    def __init__(
        self,
        db: Database,
        user_id: str,
        *,
        max_file_size: int = DEFAULT_MAX_FILE_SIZE,
    ):
        self._db = db
        self._user_id = user_id
        self._max_file_size = max_file_size

    def read(
        self, path: str, *, offset: int | None = None, limit: int | None = None
    ) -> str:
        norm = normalize(path)
        node = self._db.get_node(self._user_id, norm)
        if not node:
            raise NotFoundError(norm)
        if node["is_dir"]:
            raise IsDirectoryError(norm)
        content = node["content"] or ""
        if offset is not None or limit is not None:
            lines = content.split("\n")
            start = (offset or 1) - 1  # 1-based to 0-based
            count = limit or len(lines)
            return "\n".join(lines[max(0, start) : start + count])
        return content

    def write(self, path: str, content: str) -> None:
        norm = normalize(path)
        size = len(content.encode("utf-8"))
        if size > self._max_file_size:
            raise FileSizeError(size, self._max_file_size)

        self._ensure_parents(norm)

        existing = self._db.get_node(self._user_id, norm)
        if existing and existing["is_dir"]:
            raise IsDirectoryError(norm)
        version = existing["version"] + 1 if existing else 1

        self._db.upsert_node(
            NodeRow(
                id=existing["id"] if existing else str(uuid.uuid4()),
                user_id=self._user_id,
                path=norm,
                parent_path=parent_path(norm),
                name=base_name(norm),
                is_dir=False,
                content=content,
                version=version,
                size=size,
                created_at="",
                updated_at="",
            )
        )

    def edit(self, path: str, old_string: str, new_string: str) -> None:
        norm = normalize(path)
        node = self._db.get_node(self._user_id, norm)
        if not node:
            raise NotFoundError(norm)
        if node["is_dir"]:
            raise IsDirectoryError(norm)

        content = node["content"] or ""
        occurrences = content.count(old_string)

        if occurrences == 0:
            raise EditConflictError(f"old_string not found in {norm}")
        if occurrences > 1:
            raise EditConflictError(
                f"old_string is not unique in {norm} (found {occurrences} occurrences)"
            )

        new_content = content.replace(old_string, new_string, 1)
        new_size = len(new_content.encode("utf-8"))
        if new_size > self._max_file_size:
            raise FileSizeError(new_size, self._max_file_size)

        self._db.update_content(
            self._user_id, norm, new_content, new_size, node["version"] + 1
        )

    def multi_edit(
        self, path: str, edits: list[dict[str, str]]
    ) -> None:
        norm = normalize(path)
        node = self._db.get_node(self._user_id, norm)
        if not node:
            raise NotFoundError(norm)
        if node["is_dir"]:
            raise IsDirectoryError(norm)

        content = node["content"] or ""
        for edit in edits:
            old = edit["old_string"]
            new = edit["new_string"]
            occurrences = content.count(old)
            if occurrences == 0:
                raise EditConflictError(
                    f"old_string not found in {norm}: {old[:40]}"
                )
            if occurrences > 1:
                raise EditConflictError(
                    f"old_string is not unique in {norm} "
                    f"(found {occurrences} occurrences): {old[:40]}"
                )
            content = content.replace(old, new, 1)

        new_size = len(content.encode("utf-8"))
        if new_size > self._max_file_size:
            raise FileSizeError(new_size, self._max_file_size)

        self._db.update_content(
            self._user_id, norm, content, new_size, node["version"] + 1
        )

    def ls(
        self, path: str, *, recursive: bool = False
    ) -> list[dict]:
        norm = normalize(path)

        if norm != "/":
            node = self._db.get_node(self._user_id, norm)
            if not node:
                raise NotFoundError(norm)
            if not node["is_dir"]:
                raise NotDirectoryError(norm)

        if recursive:
            descendants = self._db.list_descendants(self._user_id, norm)
            return [
                {"path": c["path"], "name": c["name"], "is_dir": c["is_dir"], "size": c["size"]}
                for c in descendants
            ]

        children = self._db.list_children(self._user_id, norm)
        return [
            {"path": c["path"], "name": c["name"], "is_dir": c["is_dir"], "size": c["size"]}
            for c in children
        ]

    def mkdir(self, path: str) -> None:
        norm = normalize(path)
        if norm == "/":
            return
        self._ensure_parents(norm)
        self._ensure_dir(norm)

    def rm(self, path: str) -> None:
        norm = normalize(path)
        if norm == "/":
            raise ValueError("Cannot remove root directory")
        node = self._db.get_node(self._user_id, norm)
        if not node:
            raise NotFoundError(norm)

        if node["is_dir"]:
            self._db.delete_tree(self._user_id, norm)
        else:
            self._db.delete_node(self._user_id, norm)

    def append(self, path: str, content: str) -> None:
        norm = normalize(path)
        node = self._db.get_node(self._user_id, norm)
        if not node:
            self.write(norm, content)
            return
        if node["is_dir"]:
            raise IsDirectoryError(norm)
        new_content = (node["content"] or "") + content
        new_size = len(new_content.encode("utf-8"))
        if new_size > self._max_file_size:
            raise FileSizeError(new_size, self._max_file_size)

        self._db.update_content(
            self._user_id, norm, new_content, new_size, node["version"] + 1
        )

    def grep(
        self,
        pattern: str,
        path: str | None = None,
        *,
        case_insensitive: bool = False,
    ) -> list[dict]:
        if len(pattern) > 1000:
            raise ValueError("Regex pattern too long (max 1000 characters)")

        path_prefix = normalize(path) if path else None

        has_regex = bool(re.search(r"[.*+?^${}()|[\]\\]", pattern))
        like_str = "%" if has_regex else pattern
        candidates = self._db.search_content(self._user_id, like_str, path_prefix)

        flags = re.MULTILINE | (re.IGNORECASE if case_insensitive else 0)
        regex = re.compile(pattern, flags)
        results: list[dict] = []

        for node in candidates:
            if not node["content"]:
                continue
            file_lines = node["content"].split("\n")
            matched = []
            for i, line in enumerate(file_lines):
                if regex.search(line):
                    matched.append({"line": i + 1, "text": line})
            if matched:
                results.append({"path": node["path"], "lines": matched})

        return results

    def glob(
        self,
        pattern: str,
        path: str | None = None,
        *,
        type: str | None = None,
    ) -> list[str]:
        path_prefix = normalize(path) if path else None
        like_pattern = glob_to_like(pattern)
        nodes = self._db.search_names(self._user_id, like_pattern, path_prefix)
        if type == "file":
            nodes = [n for n in nodes if not n["is_dir"]]
        elif type == "dir":
            nodes = [n for n in nodes if n["is_dir"]]
        return [n["path"] for n in nodes]

    def mv(self, src: str, dst: str) -> None:
        norm_from = normalize(src)
        norm_to = normalize(dst)

        source_node = self._db.get_node(self._user_id, norm_from)
        if not source_node:
            raise NotFoundError(norm_from)

        dest_node = self._db.get_node(self._user_id, norm_to)
        final_to = norm_to
        if dest_node and dest_node["is_dir"]:
            final_to = normalize(norm_to + "/" + base_name(norm_from))

        final_target = self._db.get_node(self._user_id, final_to)
        if final_target and not final_target["is_dir"]:
            self._db.delete_node(self._user_id, final_to)

        self._ensure_parents(final_to)

        self._db.move_node(
            self._user_id,
            norm_from,
            final_to,
            parent_path(final_to),
            base_name(final_to),
        )

        if source_node["is_dir"]:
            self._db.move_tree(self._user_id, norm_from, final_to)

    def _ensure_parents(self, path: str) -> None:
        for anc in all_ancestors(path):
            self._ensure_dir(anc)

    def _ensure_dir(self, path: str) -> None:
        if path == "/":
            return
        existing = self._db.get_node(self._user_id, path)
        if existing:
            if not existing["is_dir"]:
                raise NotDirectoryError(path)
            return
        self._db.upsert_node(
            NodeRow(
                id=str(uuid.uuid4()),
                user_id=self._user_id,
                path=path,
                parent_path=parent_path(path),
                name=base_name(path),
                is_dir=True,
                content=None,
                version=1,
                size=0,
                created_at="",
                updated_at="",
            )
        )

File: ts/src/fs/operations.ts
import { randomUUID } from "node:crypto";
import type { Database, NodeRow } from "../db/types.js";
import { normalize, parentPath, baseName, allAncestors, globToLike } from "./paths.js";
import {
  NotFoundError,
  IsDirectoryError,
  NotDirectoryError,
  EditConflictError,
  FileSizeError,
} from "./errors.js";

const DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024; // 10 MB

export interface FileSystemOptions {
  /** Maximum file size in bytes. Default: 10 MB. Set to Infinity to disable. */
  maxFileSize?: number;
}

export class FileSystem {
  private maxFileSize: number;

  constructor(
    private db: Database,
    private userId: string,
    options?: FileSystemOptions
  ) {
    this.maxFileSize = options?.maxFileSize ?? DEFAULT_MAX_FILE_SIZE;
  }

  async read(path: string, opts?: { offset?: number; limit?: number }): Promise<string> {
    const norm = normalize(path);
    const node = await this.db.getNode(this.userId, norm);
    if (!node) throw new NotFoundError(norm);
    if (node.is_dir) throw new IsDirectoryError(norm);
    const content = node.content ?? "";
    if (opts?.offset !== undefined || opts?.limit !== undefined) {
      const lines = content.split("\n");
      const start = (opts.offset ?? 1) - 1; // 1-based to 0-based
      const count = opts.limit ?? lines.length;
      return lines.slice(Math.max(0, start), start + count).join("\n");
    }
    return content;
  }

  async write(path: string, content: string): Promise<void> {
    const norm = normalize(path);
    const size = Buffer.byteLength(content, "utf-8");
    if (size > this.maxFileSize) throw new FileSizeError(size, this.maxFileSize);

    await this.ensureParents(norm);

    const existing = await this.db.getNode(this.userId, norm);
    if (existing && existing.is_dir) throw new IsDirectoryError(norm);
    const version = existing ? existing.version + 1 : 1;

    await this.db.upsertNode({
      id: existing?.id ?? randomUUID(),
      user_id: this.userId,
      path: norm,
      parent_path: parentPath(norm),
      name: baseName(norm),
      is_dir: false,
      content,
      version,
      size,
    });
  }

  async edit(path: string, oldString: string, newString: string): Promise<void> {
    const norm = normalize(path);
    const node = await this.db.getNode(this.userId, norm);
    if (!node) throw new NotFoundError(norm);
    if (node.is_dir) throw new IsDirectoryError(norm);

    const content = node.content ?? "";
    const occurrences = content.split(oldString).length - 1;

    if (occurrences === 0) {
      throw new EditConflictError(
        `old_string not found in ${norm}`
      );
    }
    if (occurrences > 1) {
      throw new EditConflictError(
        `old_string is not unique in ${norm} (found ${occurrences} occurrences)`
      );
    }

    const newContent = content.replace(oldString, newString);
    const newSize = Buffer.byteLength(newContent, "utf-8");
    if (newSize > this.maxFileSize) throw new FileSizeError(newSize, this.maxFileSize);

    await this.db.updateContent(
      this.userId,
      norm,
      newContent,
      newSize,
      node.version + 1
    );
  }

  async multiEdit(path: string, edits: Array<{ old_string: string; new_string: string }>): Promise<void> {
    const norm = normalize(path);
    const node = await this.db.getNode(this.userId, norm);
    if (!node) throw new NotFoundError(norm);
    if (node.is_dir) throw new IsDirectoryError(norm);

    let content = node.content ?? "";
    for (const edit of edits) {
      const occurrences = content.split(edit.old_string).length - 1;
      if (occurrences === 0) {
        throw new EditConflictError(`old_string not found in ${norm}: ${edit.old_string.slice(0, 40)}`);
      }
      if (occurrences > 1) {
        throw new EditConflictError(
          `old_string is not unique in ${norm} (found ${occurrences} occurrences): ${edit.old_string.slice(0, 40)}`
        );
      }
      content = content.replace(edit.old_string, edit.new_string);
    }

    const newSize = Buffer.byteLength(content, "utf-8");
    if (newSize > this.maxFileSize) throw new FileSizeError(newSize, this.maxFileSize);

    await this.db.updateContent(
      this.userId,
      norm,
      content,
      newSize,
      node.version + 1
    );
  }

  async ls(path: string, opts?: { recursive?: boolean }): Promise<Array<{ path: string; name: string; isDir: boolean; size: number }>> {
    const norm = normalize(path);

    // Root always exists
    if (norm !== "/") {
      const node = await this.db.getNode(this.userId, norm);
      if (!node) throw new NotFoundError(norm);
      if (!node.is_dir) throw new NotDirectoryError(norm);
    }

    if (opts?.recursive) {
      const descendants = await this.db.listDescendants(this.userId, norm);
      return descendants.map((c) => ({
        path: c.path,
        name: c.name,
        isDir: c.is_dir,
        size: c.size,
      }));
    }

    const children = await this.db.listChildren(this.userId, norm);
    return children.map((c) => ({
      path: c.path,
      name: c.name,
      isDir: c.is_dir,
      size: c.size,
    }));
  }

  async mkdir(path: string): Promise<void> {
    const norm = normalize(path);
    if (norm === "/") return;
    await this.ensureParents(norm);
    await this.ensureDir(norm);
  }

  async rm(path: string): Promise<void> {
    const norm = normalize(path);
    if (norm === "/") {
      throw new Error("Cannot remove root directory");
    }
    const node = await this.db.getNode(this.userId, norm);
    if (!node) throw new NotFoundError(norm);

    if (node.is_dir) {
      await this.db.deleteTree(this.userId, norm);
    } else {
      await this.db.deleteNode(this.userId, norm);
    }
  }

  async append(path: string, content: string): Promise<void> {
    const norm = normalize(path);
    const node = await this.db.getNode(this.userId, norm);
    if (!node) {
      // File doesn't exist — create it
      await this.write(norm, content);
      return;
    }
    if (node.is_dir) throw new IsDirectoryError(norm);
    const newContent = (node.content ?? "") + content;
    const newSize = Buffer.byteLength(newContent, "utf-8");
    if (newSize > this.maxFileSize) throw new FileSizeError(newSize, this.maxFileSize);

    await this.db.updateContent(
      this.userId,
      norm,
      newContent,
      newSize,
      node.version + 1
    );
  }

  async grep(
    pattern: string,
    path?: string,
    opts?: { case_insensitive?: boolean }
  ): Promise<Array<{ path: string; lines: Array<{ line: number; text: string }> }>> {
    if (pattern.length > 1000) {
      throw new Error("Regex pattern too long (max 1000 characters)");
    }

    const pathPrefix = path ? normalize(path) : undefined;

    // If pattern contains regex syntax, scan all files (DB can't pre-filter).
    // Only use LIKE hint when the pattern is plain literal text.
    const hasRegexSyntax = /[.*+?^${}()|[\]\\]/.test(pattern);
    const likeStr = hasRegexSyntax ? "%" : pattern;
    const candidates = await this.db.searchContent(
      this.userId,
      likeStr,
      pathPrefix
    );

    const flags = "gm" + (opts?.case_insensitive ? "i" : "");
    const regex = new RegExp(pattern, flags);
    const results: Array<{ path: string; lines: Array<{ line: number; text: string }> }> = [];

    for (const node of candidates) {
      if (!node.content) continue;
      const fileLines = node.content.split("\n");
      const matched: Array<{ line: number; text: string }> = [];
      for (let i = 0; i < fileLines.length; i++) {
        regex.lastIndex = 0;
        if (regex.test(fileLines[i])) {
          matched.push({ line: i + 1, text: fileLines[i] });
        }
      }
      if (matched.length > 0) {
        results.push({ path: node.path, lines: matched });
      }
    }

    return results;
  }

  async glob(
    pattern: string,
    path?: string,
    opts?: { type?: "file" | "dir" }
  ): Promise<string[]> {
    const pathPrefix = path ? normalize(path) : undefined;
    const likePattern = globToLike(pattern);
    let nodes = await this.db.searchNames(this.userId, likePattern, pathPrefix);
    if (opts?.type === "file") {
      nodes = nodes.filter((n) => !n.is_dir);
    } else if (opts?.type === "dir") {
      nodes = nodes.filter((n) => n.is_dir);
    }
    return nodes.map((n) => n.path);
  }

  async mv(from: string, to: string): Promise<void> {
    const normFrom = normalize(from);
    const normTo = normalize(to);

    const sourceNode = await this.db.getNode(this.userId, normFrom);
    if (!sourceNode) throw new NotFoundError(normFrom);

    // Check if destination is an existing directory — move into it
    const destNode = await this.db.getNode(this.userId, normTo);
    let finalTo = normTo;
    if (destNode && destNode.is_dir) {
      finalTo = normalize(normTo + "/" + baseName(normFrom));
    }

    // If target is an existing file, remove it first (overwrite like real mv)
    const finalTarget = await this.db.getNode(this.userId, finalTo);
    if (finalTarget && !finalTarget.is_dir) {
      await this.db.deleteNode(this.userId, finalTo);
    }

    await this.ensureParents(finalTo);

    // Move the node itself
    await this.db.moveNode(
      this.userId,
      normFrom,
      finalTo,
      parentPath(finalTo),
      baseName(finalTo)
    );

    // If it's a directory, move all descendants
    if (sourceNode.is_dir) {
      await this.db.moveTree(this.userId, normFrom, finalTo);
    }
  }

  private async ensureParents(path: string): Promise<void> {
    const ancestors = allAncestors(path);
    for (const anc of ancestors) {
      await this.ensureDir(anc);
    }
  }

  private async ensureDir(path: string): Promise<void> {
    if (path === "/") return; // root is virtual, never stored
    const existing = await this.db.getNode(this.userId, path);
    if (existing) {
      if (!existing.is_dir) {
        throw new NotDirectoryError(path);
      }
      return;
    }
    await this.db.upsertNode({
      id: randomUUID(),
      user_id: this.userId,
      path,
      parent_path: parentPath(path),
      name: baseName(path),
      is_dir: true,
      content: null,
      version: 1,
      size: 0,
    });
  }
}
